File: src/cpp/asynchrony/06_data_sharing_queue.cpp
/*
This C++ program demonstrates data sharing and communication between asynchronous tasks
using the C++ standard library. In this example, we simulate a producer and consumer
scenario where one task produces data and the other consumes it using std::promise and
std::future for communication and synchronization.
*/

#include <iostream>
#include <thread>
#include <future>
#include <chrono>
#include <random>

int produce_data() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distr(1, 100);
    int data = distr(gen);
    std::cout << "Produced data: " << data << std::endl;
    return data;
}

void consume_data(int data) {
    std::cout << "Consumed data: " << data << std::endl;
}

int main() {
    std::promise<int> data_promise;
    std::future<int> data_future = data_promise.get_future();

    // Producer task
    auto producer_task = std::async(std::launch::async, [&data_promise]() {
        int data = produce_data();
        data_promise.set_value(data);
    });

    // Consumer task
    auto consumer_task = std::async(std::launch::async, [&data_future]() {
        int data = data_future.get();
        consume_data(data);
    });

    producer_task.wait();
    consumer_task.wait();

    return 0;
}

File: src/python/asynchrony/06_data_sharing_queue.py
"""
This script demonstrates the use of asyncio.Queue for data sharing and communication
between async coroutines in Python. An asyncio.Queue is a simple way to send and receive
messages between coroutines in an asynchronous manner, allowing them to communicate
and synchronize their operations without the need for locks or other synchronization
primitives. In this example, we simulate a producer and consumer scenario where one
coroutine produces data and the other consumes it using an asyncio.Queue.
"""

import asyncio
import random


async def producer(queue):
    for i in range(10):
        await asyncio.sleep(random.uniform(0.5, 1))
        print(f"Producer: Produced item {i}")
        await queue.put(i)
    await queue.put(None)  # Signal the consumer to exit by sending None


async def consumer(queue):
    while True:
        item = await queue.get()
        if item is None:
            break
        print(f"Consumer: Consumed item {item}")
        await asyncio.sleep(random.uniform(0.5, 1))


async def main():
    queue = asyncio.Queue()
    producer_task = asyncio.create_task(producer(queue))
    consumer_task = asyncio.create_task(consumer(queue))
    await asyncio.gather(producer_task, consumer_task)


if __name__ == "__main__":
    asyncio.run(main())
